src/decode/lzma.rs

```diff
@@ -151,136 +144,86 @@ where
     rep_len_decoder: rangecoder::LenDecoder,
 }
 
-// Initialize decoder with accumulating buffer
-pub fn new_accum<W>(
-    output: lzbuffer::LzAccumBuffer<W>,
-    lc: u32,
-    lp: u32,
-    pb: u32,
-    unpacked_size: Option<u64>,
-) -> DecoderState<W, lzbuffer::LzAccumBuffer<W>>
-where
-    W: io::Write,
-{
-    DecoderState {
-        _phantom: PhantomData,
-        partial_input_buf: std::io::Cursor::new([0; MAX_REQUIRED_INPUT]),
-        output,
-        lc,
-        lp,
-        pb,
-        unpacked_size,
-        literal_probs: vec![vec![0x400; 0x300]; 1 << (lc + lp)],
-        pos_slot_decoder: vec![rangecoder::BitTree::new(6); 4],
-        align_decoder: rangecoder::BitTree::new(4),
-        pos_decoders: [0x400; 115],
-        is_match: [0x400; 192],
-        is_rep: [0x400; 12],
-        is_rep_g0: [0x400; 12],
-        is_rep_g1: [0x400; 12],
-        is_rep_g2: [0x400; 12],
-        is_rep_0long: [0x400; 192],
-        state: 0,
-        rep: [0; 4],
-        len_decoder: rangecoder::LenDecoder::new(),
-        rep_len_decoder: rangecoder::LenDecoder::new(),
+impl DecoderState {
+    // Initialize decoder with accumulating buffer
+    pub fn new(lc: u32, lp: u32, pb: u32, unpacked_size: Option<u64>) -> Self {
+        assert!(lc < 8);
+        assert!(lp < 4);
+        assert!(pb < 4);
+
+        DecoderState {
+            partial_input_buf: std::io::Cursor::new([0; MAX_REQUIRED_INPUT]),
+            lc,
+            lp,
+            pb,
+            unpacked_size,
+            literal_probs: vec![vec![0x400; 0x300]; 1 << (lc + lp)],
+            pos_slot_decoder: vec![rangecoder::BitTree::new(6); 4],
+            align_decoder: rangecoder::BitTree::new(4),
+            pos_decoders: [0x400; 115],
+            is_match: [0x400; 192],
+            is_rep: [0x400; 12],
+            is_rep_g0: [0x400; 12],
+            is_rep_g1: [0x400; 12],
+            is_rep_g2: [0x400; 12],
+            is_rep_0long: [0x400; 192],
+            state: 0,
+            rep: [0; 4],
+            len_decoder: rangecoder::LenDecoder::new(),
+            rep_len_decoder: rangecoder::LenDecoder::new(),
+        }
     }
-}
 
-// Initialize decoder with circular buffer
-pub fn new_circular<W>(
-    output: W,
-    params: LzmaParams,
-) -> error::Result<DecoderState<W, lzbuffer::LzCircularBuffer<W>>>
-where
-    W: io::Write,
-{
-    new_circular_with_memlimit(output, params, std::usize::MAX)
-}
+    pub fn reset_state(&mut self, lc: u32, lp: u32, pb: u32) {
+        assert!(lc < 8);
+        assert!(lp < 4);
+        assert!(pb < 4);
```

Comment:
I now realized that the ranges in the documentation are inconsistent with their actual definition in `LzmaParams` above:

```rust
        let lc: u32 = pb % 9;
        pb /= 9;
        let lp: u32 = pb % 5;
        pb /= 5;
```

Can you update the assertions, and also the docs to e.g. `0..=8`?

Ideally, the boundary condition on these values is also something we'd want to test (either as part of #33, or as new test cases - and also comparing to other implementations i.e. #20).

Reply:
Fixed this and moved assertions to `LzmaProperties::validate`